mlir/lib/Dialect/DLTI/DLTI.cpp

```diff
@@ -438,6 +446,15 @@ FailureOr<Attribute> dlti::query(Operation *op, ArrayRef<StringAttr> keys,
     return failure();
   }
 
+  auto keyToStr = [](DataLayoutEntryKey key) -> std::string {
+    std::string buf;
+    llvm::TypeSwitch<DataLayoutEntryKey>(key)
+        .Case<StringAttr, Type>( // The only two kinds of key we know of.
+            [&](auto key) { llvm::raw_string_ostream(buf) << key; })
+        .Default([](auto) { llvm_unreachable("unexpected entry key kind"); });
+    return buf;
+  };
```

Comment:
That seems expensive to me for the purpose of a query!

Why can't we use a `void*`?

Reply:
This lambda is only invoked when `emitError=true`, i.e. these conversions to a string only happen when error messages get generated.

I am only making use of `::print(..)` here because I had no other way of obtaining an accurate representation of a `Type` key. If there's a more efficient way to get a string representation of an arbitrary `Type`, I am more than happy to use that instead.

Reply:
> This lambda is only invoked when emitError=true, i.e. these conversions to a string only happen when error messages get generated and solely in service of these error messages.

Oh alright then!